akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectivesSpec.scala

```diff
@@ -19,108 +21,136 @@ import akka.http.scaladsl.model.MediaTypes._
+        mediaType.withParams(Map.empty) shouldEqual `multipart/byteranges`
 
-          val parts = responseAs[Multipart.ByteRanges].toStrict(1.second).awaitResult(3.seconds).strictParts
-          parts.map(_.entity.data.utf8String) should contain theSameElementsAs List("BCDEFGHIJK", "QRSTUVWXYZ")
-        }
-      } finally file.delete
+        val parts = responseAs[Multipart.ByteRanges].toStrict(1.second).awaitResult(3.seconds).strictParts
+        parts.map(_.entity.data.utf8String) should contain theSameElementsAs List("BCDEFGHIJK", "QRSTUVWXYZ")
+      }
     }
 
     "properly handle zero-byte files" in {
-      val file = File.createTempFile("akkaHttpTest", null)
-      try {
-        Get() ~> getFromFile(file) ~> check {
-          mediaType shouldEqual NoMediaType
-          responseAs[String] shouldEqual ""
-        }
-      } finally file.delete
+      Get() ~> _getFromFile("empty") ~> check {
+        mediaType shouldEqual NoMediaType
+        responseAs[String] shouldEqual ""
+      }
     }
 
     "support precompressed files with registered MediaType" in {
-      val file = File.createTempFile("akkaHttpTest", ".svgz")
-      try {
-        writeAllText("123", file)
-        Get() ~> getFromFile(file) ~> check {
-          mediaType shouldEqual `image/svg+xml`
-          header[`Content-Encoding`] shouldEqual Some(`Content-Encoding`(HttpEncodings.gzip))
-          responseAs[String] shouldEqual "123"
-        }
-      } finally file.delete
+      Get() ~> _getFromFile("sample.svgz") ~> check {
+        mediaType shouldEqual `image/svg+xml`
+        header[`Content-Encoding`] shouldEqual Some(`Content-Encoding`(HttpEncodings.gzip))
+        responseAs[String] shouldEqual "123"
+      }
     }
 
     "support files with registered MediaType and .gz suffix" in {
-      val file = File.createTempFile("akkaHttpTest", ".js.gz")
+      Get() ~> _getFromFile("sample.js.gz") ~> check {
+        mediaType shouldEqual `application/javascript`
+        header[`Content-Encoding`] shouldEqual Some(`Content-Encoding`(HttpEncodings.gzip))
+        responseAs[String] shouldEqual "456"
+      }
+    }
+  }
+
+  "getFromDirectory" should {
+    def getFromDir(directory: String) = getFromDirectory(base.resolve(directory))
+
+    "reject non-GET requests" in {
+      Put() ~> getFromDir("someDir") ~> check { handled shouldEqual false }
+    }
+    "reject requests to non-existing files" in {
+      Get("nonExistentFile") ~> getFromDir("subDirectory") ~> check { handled shouldEqual false }
+    }
+    "reject requests to directories" in {
+      Get("sub") ~> getFromDir("someDir") ~> check { handled shouldEqual false }
+    }
+    "reject path traversal attempts" in {
+      val route = getFromDir("someDir/sub")
+      Get("file.html") ~> route ~> check { handled shouldEqual true }
+      def shouldReject(prefix: String) = Get(prefix + "fileA.txt") ~> route ~> check { handled shouldEqual false }
+      shouldReject("../")
```

Comment:
BTW: `RouteTest` does `request.withEffectiveUri` which ends in `URI#collapseDotSegments` which in turn cuts off the dots. Will have to find a way around that for proper tests.